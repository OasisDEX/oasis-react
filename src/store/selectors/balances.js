import { createSelector } from "reselect";
import BigNumber from "bignumber.js";
import reselect from "../../utils/reselect";
import web3 from "../../bootstrap/web3";
import {
  ETH_UNIT_ETHER,
  TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED,
  TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED,
  TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED_MIN
} from "../../constants";

import tokens from "./tokens";
import { getMarketContractInstance } from "../../bootstrap/contracts";

const balances = s => s.get("balances");

const tokenAllowances = createSelector(balances, s => s.get("tokenAllowances"));

const tokenBalances = createSelector(balances, s => s.get("tokenBalances"));

const tokenBalance = createSelector(
  tokenBalances,
  reselect.getProps,
  (s, { tokenName, balanceUnit = ETH_UNIT_ETHER, toBigNumber = true }) => {
    const tokenBalance = s.get(tokenName);
    if (tokenBalance) {
      if (toBigNumber) {
        return web3.fromWei(new BigNumber(s.get(tokenName), 10), balanceUnit);
      } else {
        return web3.fromWei(s.get(tokenName), balanceUnit);
      }
    } else {
      return null;
    }
  }
);

const tokenAllowanceTrustStatus = createSelector(
  balances,
  reselect.getProps,
  (s, { tokenName, allowanceSubjectAddress }) => {
    const tokenAllowance = s.getIn([
      "tokenAllowances",
      tokenName,
      allowanceSubjectAddress
    ]);
    if (tokenAllowance) {
      const tokenAllowanceBN = new BigNumber(tokenAllowance);
      const tokenTrustEnabledMinBN = new BigNumber(
        TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED_MIN
      );
      if (tokenAllowanceBN.gte(tokenTrustEnabledMinBN)) {
        return TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED;
      } else {
        return TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED;
      }
    } else {
      return null;
    }
  }
);


const tokenAllowanceStatusForActiveMarket = createSelector(
  (...args) => args,
  ([state, { tokenName }]) =>
    tokenAllowanceTrustStatus(state, {
      tokenName,
      allowanceSubjectAddress: getMarketContractInstance().address
    })
);

const activeBaseTokenBalance = createSelector(
  tokenBalances,
  tokens.activeTradingPairBaseToken,
  (tokenBalances, baseToken) => tokenBalances.get(baseToken)
);

const ethBalance = createSelector(balances, s => s.get("ethBalance"));

const activeQuoteTokenBalance = createSelector(
  tokenBalances,
  tokens.activeTradingPairQuoteToken,
  (tokenBalances, quoteToken) => tokenBalances.get(quoteToken)
);

export default {
  state: balances,
  tokenAllowances,
  tokenBalances,
  // tokenAllowance,
  tokenBalance,
  ethBalance,
  tokenAllowanceTrustStatus,
  activeBaseTokenBalance,
  activeQuoteTokenBalance,
  tokenAllowanceStatusForActiveMarket,
};
