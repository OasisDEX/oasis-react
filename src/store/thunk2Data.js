export default () =>  {
  let first = true;
  return store => next => action => {

    if(first) {
      first = false;
      return next(action);
    }

    if (action instanceof Array) {
      const [thunk, ...args] = action;
      return next({type: 'deferred-thunk', payload: {thunk, thunkName: thunk.name, args}});
    } else if (typeof action === 'function') {
      return next({type: 'thunk', payload: {thunk: action, thunkName: action.name}});
    } else {
      return next(action);
    }
  }
}