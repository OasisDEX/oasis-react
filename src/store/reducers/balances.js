import { createAction, handleActions } from 'redux-actions';
import Immutable from 'immutable';
import {createPromiseActions} from '../../utils/createPromiseActions';
import {fulfilled} from '../../utils/store';


const initialState = Immutable.fromJS({
  accounts: [],
  defaultAccount: {
    address: null,
    ethBalance: null,
    tokenBalances: {},
    tokenAllowances: {}
  },

});

const INIT = 'BALANCES/INIT';
const GET_DEFAULT_ACCOUNT_ETH_BALANCE = 'BALANACES/GET_DEFAULT_ACCOUNT_ETH_BALANCE';
const GET_ALL_TRADED_TOKENS_BALANCES = 'BALANACES/GET_ALL_TRADED_TOKENS_BALANCES';
const GET_ALL_TRADED_TOKENS_ALLOWANCES = 'BALANACES/GET_ALL_TRADED_TOKENS_ALLOWANCES';

const Init = createAction(
  INIT,
  () => null,
);


const getDefaultAccountEthBalance = createAction(
  GET_DEFAULT_ACCOUNT_ETH_BALANCE,
  async () => window.web3p.eth.getBalance(window.web3.eth.defaultAccount)
);



const getAllTradedTokensBalances = createAction(
  GET_ALL_TRADED_TOKENS_BALANCES,
  async (tokensContractsLists) => {

    const tokensBalancesPromises = [];

    Object.entries(tokensContractsLists).forEach(
      async ([tokenName, tokenContract]) => {
          if(tokenContract.balanceOf) {
          tokensBalancesPromises.push(
            tokenContract.balanceOf(window.web3.eth.defaultAccount)
          );
        }
      }
    );

    return Promise.all(tokensBalancesPromises).then(
      tokenBalances => {

        const balancesByToken = {};
        Object.keys(tokensContractsLists).forEach(
          (tokenName, i) => balancesByToken[tokenName] = tokenBalances[i]
        );

        return balancesByToken;
      }
    )
  }
);

const getAllTradedTokensAllowances = createAction(
  GET_ALL_TRADED_TOKENS_ALLOWANCES,
  async (tokensContractsLists, spenderAddress) => {

    const tokensAllowancesPromises = [];

    Object.entries(tokensContractsLists).forEach(
      async ([tokenName, tokenContract]) => {
        if(tokenContract.allowance) {
          tokensAllowancesPromises.push(
            tokenContract.allowance(window.web3.eth.defaultAccount, spenderAddress)
          );
        }
      }
    );

    return Promise.all(tokensAllowancesPromises).then(
      tokenAllowances => {
        const allowancesByToken = {};
        Object.keys(tokensContractsLists).forEach(
          (tokenName, i) => allowancesByToken[tokenName] = tokenAllowances[i]
        );

        return allowancesByToken;
      }
    )
  }
);



const tokenTransferEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER',
  (tokenName, userAddress, event) => ({ tokenName, event })
)

const tokenTransferFromEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER_FROM',
  (tokenName, userAddress, event) => ({ tokenName, userAddress, event })
)

const tokenTransferToEvent = createAction(
  'BALANCES/EVENT___TOKEN_TRANSFER_TO',
  (tokenName, userAddress, event) => ({ tokenName, userAddress, event })
);

const subscribeTokenTransfersEvents = createPromiseActions(
  'BALANCES/SUBSCRIBE_TOKEN_TRANSFER_EVENT'
);

const subscribeTokenTransfersEventsEpic = (tokensContractsList, address, config = {}) => async (dispatch) => {
  dispatch(subscribeTokenTransfersEvents.pending());
  Object.entries(tokensContractsList).forEach(
    ([tokenName, tokenContract]) => {
      /**
       * Listen to all erc20 transfer events from now.
       */
      tokenContract.Transfer(address, config)
        .then(
          transferEvent => {
            dispatch(tokenTransferEvent(tokenName, transferEvent, address));
          }
        );
    }
  );
  dispatch(subscribeTokenTransfersEvents.fulfilled());
};

const actions = {
  Init,
  getDefaultAccountEthBalance,
  getAllTradedTokensBalances,
  getAllTradedTokensAllowances,
  subscribeTokenTransfersEventsEpic
};

const reducer = handleActions({

  [fulfilled(getAllTradedTokensBalances)]: (state, action ) =>
    state.updateIn(
      ['defaultAccount', 'tokenBalances'],
      (balances) => {
        const tokenBalances = action.payload;
        Object.entries(tokenBalances).forEach(
          ([tokenName, tokenBalance]) => {
            balances = balances.set(tokenName, tokenBalance ? tokenBalance.toFormat(10) : null)
          }
        );
        return balances
      }),

  [fulfilled(getAllTradedTokensAllowances)]: (state, action ) =>
    state.updateIn(
      ['defaultAccount', 'tokenAllowances'],
      (allowances) => {
        const tokenAllowances = action.payload;
        Object.entries(tokenAllowances).forEach(
          ([tokenName, tokenAllowance]) => {
            allowances = allowances.set(tokenName, tokenAllowance.toFormat(10))
          }
        );
        return allowances
      }),
  [tokenTransferEvent]: (state, { payload: { tokenName, userAddress, event } }) => {

    console.log({tokenName, userAddress, event})
   return state.updateIn(['tokenBalances', tokenName], (balance) => balance)
  }


}, initialState);

export default {
  actions,
  reducer,
};
