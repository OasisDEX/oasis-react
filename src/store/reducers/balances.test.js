import thunk from 'redux-thunk';
import thunk2Data from '../thunk2Data';
import configureMockStore from 'redux-mock-store'

import { fromJS } from 'immutable';

import {
  ETH_UNIT_ETHER,
  ETH_UNIT_WEI,
  TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED,
  TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED_MIN_MAX,
  TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED,
} from '../../constants';

import balances from './balances'

import each from 'jest-each';

const testCases = [
  ['MKR', TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED, TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED, 0, false],
  ['MKR', TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED, TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED, 1, false],
  ['MKR', TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED, TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED, 0, false],
  ['MKR', TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED, TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED, 1, false],
  ['DAI', TOKEN_ALLOWANCE_TRUST_STATUS_DISABLED, TOKEN_ALLOWANCE_TRUST_STATUS_ENABLED, 1, true],
]

each(testCases).describe(
  "setTokenAllowanceTrustEpic",
  (tokenName, previousAllowance, newAllowance, expectedNoCalls, expectedToTrow) =>
{

  test(`token: ${tokenName}, status: ${previousAllowance} set to: ${newAllowance}`, () => {
    const store = configureMockStore([thunk2Data(), thunk])({});

    const handleTransactionMock = jest.fn();

    const meat = () => store.dispatch(balances.actions.setTokenAllowanceTrustEpic({
        tokenName: tokenName,
        newAllowanceTrustStatus: newAllowance,
        allowanceSubjectAddress: 'mock address'
      },
      {},
      {
        defaultAccount: () => '',
        tokenAllowanceTrustStatus: () => previousAllowance,
        getErc20Tokens: () => fromJS(['MKR']),
        handleTrans: handleTransactionMock
      }));

    if(expectedToTrow) {
      expect(meat).toThrow();
    } else {
      meat();

      expect(store.getActions()).toMatchSnapshot();

      expect(handleTransactionMock.mock.calls.length).toBe(expectedNoCalls);

      if (expectedNoCalls > 0) {
        expect(handleTransactionMock.mock.calls[0][0]).toMatchSnapshot();
      }
    }
  });
});
