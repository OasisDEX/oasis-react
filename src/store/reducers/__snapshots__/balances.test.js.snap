// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setTokenAllowanceTrustEpic token: MKR, status: false set to: false 1`] = `
Array [
  Object {
    "type": "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_PENDING",
  },
  Object {
    "payload": "Trust status did not change",
    "type": Array [
      "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_REJECTED",
    ],
  },
]
`;

exports[`setTokenAllowanceTrustEpic token: MKR, status: false set to: true 1`] = `
Array [
  Object {
    "type": "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_PENDING",
  },
]
`;

exports[`setTokenAllowanceTrustEpic token: MKR, status: false set to: true 2`] = `
Object {
  "dispatch": [Function],
  "onTransactionCompleted": [Function],
  "transactionDispatcher": [Function],
  "transactionType": "TX/GROUP__LIMITS/TX/ALLOWANCE_TRUST_TOGGLE",
  "txMeta": Object {
    "account": "mock address",
    "tokenName": "MKR",
  },
  "withCallbacks": Object {},
}
`;

exports[`setTokenAllowanceTrustEpic token: MKR, status: true set to: false 1`] = `
Array [
  Object {
    "type": "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_PENDING",
  },
]
`;

exports[`setTokenAllowanceTrustEpic token: MKR, status: true set to: false 2`] = `
Object {
  "dispatch": [Function],
  "onTransactionCompleted": [Function],
  "transactionDispatcher": [Function],
  "transactionType": "TX/GROUP__LIMITS/TX/ALLOWANCE_TRUST_TOGGLE",
  "txMeta": Object {
    "account": "mock address",
    "tokenName": "MKR",
  },
  "withCallbacks": Object {},
}
`;

exports[`setTokenAllowanceTrustEpic token: MKR, status: true set to: true 1`] = `
Array [
  Object {
    "type": "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_PENDING",
  },
  Object {
    "payload": "Trust status did not change",
    "type": Array [
      "BALANCES/SET_TOKEN_ALLOWANCE_TRUST_STATUS_REJECTED",
    ],
  },
]
`;
